tools/calyx-writer/lib.rs

```diff
@@ -0,0 +1,1077 @@
+impl CalyxWriter for Assignment {
+    fn write(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result {
+        write!(f, "{} = ", self.lhs)?;
+        if self.guard != Guard::None {
+            self.guard.write(f)?;
+            write!(f, " ? ")?;
+        }
+        write!(f, "{};", self.rhs)
+    }
+}
+
+/// See `calyx_ir::Group`. Contains optional documentation.
+#[derive(PartialEq, Eq, Debug)]
+pub struct Group {
+    name: String,
+    description: Option<String>,
+    is_comb: bool,
+    /// `None` if dynamic, `Some(latency)` if static and taking `latency cycles.`
+    latency: Option<usize>,
+    assignments: Vec<Assignment>,
+}
+
+impl Group {
+    /// Constructs an empty group named `name`, combinational if and only if
+    /// `is_comb`.
+    fn new<S: ToString>(name: S, is_comb: bool) -> Self {
+        Self {
+            name: name.to_string(),
+            description: None,
+            is_comb,
+            latency: None,
+            assignments: vec![],
+        }
+    }
+
+    /// Sets a human-readable description of the group. This description may
+    /// take multiple lines.
+    pub fn describe(&mut self, description: String) {
+        self.description = Some(description);
+    }
+
+    pub fn set_latency(&mut self, latency: usize) {
+        self.latency = Some(latency);
+    }
+
+    pub fn is_comb(&self) -> bool {
+        self.is_comb
+    }
+
+    /// Adds an unguarded assignment between `lhs` and `rhs`. Behaves like
+    /// [`Group::assign_guarded`] with [`Guard::None`] passed for the guard.
+    pub fn assign(&mut self, lhs: Port, rhs: Port) {
+        self.assign_guarded(lhs, rhs, Guard::None);
+    }
+
+    /// Adds an assignment guarded by `guard` between `lhs` and `rhs`.
+    pub fn assign_guarded(&mut self, lhs: Port, rhs: Port, guard: Guard) {
+        self.assignments.push(Assignment::new(lhs, rhs, guard));
+    }
+}
+
+impl PortProvider for Group {
+    fn get(&self, port: String) -> Port {
+        Port::inferred(Some(self.name.clone()), port)
+    }
+}
+
+impl CalyxWriter for Group {
+    fn write(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result {
+        if let Some(description) = &self.description {
+            for line in description.lines() {
+                writeln!(f, "// {}", line)?;
+            }
+        }
+        if let Some(latency) = self.latency {
+            write!(f, "static<{}> ", latency)?;
+        }
+        if self.is_comb {
+            write!(f, "comb ")?;
+        }
+        writeln!(f, "group {} {{", self.name)?;
+        f.push();
+        for assignment in &self.assignments {
+            assignment.writeln(f)?;
+        }
+        f.pop();
+        write!(f, "}}")
+    }
+}
+
+/// Helper variant with [`Control`].
+#[derive(PartialEq, Eq, Debug)]
+enum ControlValue {
+    Empty,
+    Enable(RRC<Group>),
+    Seq(Vec<Control>),
+    Par(Vec<Control>),
+    While(Port, Option<RRC<Group>>, Vec<Control>),
+    If(Port, Option<RRC<Group>>, Vec<Control>, Vec<Control>),
```

Comment:
The bodies/true-false arms should just be `Control` not `Vec<Control>` since a
`Seq` is how you would get multiple statemens. I assume this is for ease of use?

Reply:
I thought that this was valid calyx syntax? I can change it though